File: src/lib/approvals.ts
import { invoke } from "@tauri-apps/api/core";

export type ApprovalRuleDecision = "approve" | "deny";

/**
 * Codex-style **approval policy** — the orthogonal "when do we pause to ask the
 * user?" axis that sits alongside the sandbox tier. Mirrors `ApprovalPolicy` in
 * `src-tauri/src/orchestrator/approval_policy.rs`.
 *
 *  - `untrusted`  — ask for everything except provably read-only inspection
 *                   (read tools + read-only shell commands).
 *  - `on-request` — DEFAULT. Forward every approval request to the user.
 *  - `never`      — never pause; auto-approve anything that already passed the
 *                   sandbox tier + guardrails ("full-auto within the sandbox").
 *
 * The policy only ever *skips* a prompt — it cannot widen what the sandbox tier
 * or guardrails already forbid (those gates run first in `chat.rs`).
 */
export type ApprovalPolicy = "untrusted" | "on-request" | "never";

export const APPROVAL_POLICIES: readonly ApprovalPolicy[] = [
  "untrusted",
  "on-request",
  "never",
] as const;

/** Default matches the Rust default — historic behavior (always ask). */
export const DEFAULT_APPROVAL_POLICY: ApprovalPolicy = "on-request";

function isPolicy(s: string): s is ApprovalPolicy {
  return (APPROVAL_POLICIES as readonly string[]).includes(s);
}

/** Human label + short description, for use in popovers and settings. */
export const APPROVAL_POLICY_META: Record<
  ApprovalPolicy,
  { label: string; description: string }
> = {
  untrusted: {
    label: "Untrusted",
    description:
      "Ask before anything except provably read-only inspection (git status, ls, grep…).",
  },
  "on-request": {
    label: "On request",
    description: "Ask for every tool call that needs approval (default).",
  },
  never: {
    label: "Never ask",
    description:
      "Auto-approve everything the sandbox tier + guardrails already allow.",
  },
};

/** Read the configured approval policy for `projectRoot`. Returns the default
 *  when no `.cortex/approval-policy.toml` exists or it fails to parse. */
export async function getApprovalPolicy(
  projectRoot: string,
): Promise<ApprovalPolicy> {
  const raw = await invoke<string>("get_approval_policy", { projectRoot });
  return isPolicy(raw) ? raw : DEFAULT_APPROVAL_POLICY;
}

/** Persist the approval policy for `projectRoot`. The Rust side validates too. */
export async function setApprovalPolicy(
  projectRoot: string,
  policy: ApprovalPolicy,
): Promise<void> {
  if (!isPolicy(policy)) {
    throw new Error(`invalid approval policy '${policy}'`);
  }
  return invoke("set_approval_policy", { projectRoot, policy });
}

/**
 * Append a persistent approval rule to `<projectRoot>/.cortex/approvals.toml`.
 * The backend validates `pattern` as a regex; an invalid regex bubbles up as
 * a rejected promise.
 */
export async function addApprovalRule(
  projectRoot: string,
  pattern: string,
  decision: ApprovalRuleDecision,
): Promise<void> {
  return invoke("add_approval_rule", {
    projectRoot,
    pattern,
    decision,
  });
}

/**
 * Entry in the user-global auto-approve allowlist
 * (`~/.cortex/auto-approve.json`).
 *
 *  - `tool`:    case-insensitive match against the tool name; `""` is a wildcard
 *  - `pattern`: glob (`globset` crate semantics on the backend) matched
 *               against the tool call's primary string field (`command` /
 *               `cmd` / `shell` / `bash` / `path` / `file`) or the whole
 *               serialized payload when none of those fields are present
 *  - `profile`: optional tag — surfaced in the UI but not yet enforced
 */
export interface AutoApproveEntry {
  tool: string;
  pattern: string;
  profile?: string;
}

/** Read the on-disk allowlist. Missing file → `[]`. */
export async function listAutoApprove(): Promise<AutoApproveEntry[]> {
  return invoke<AutoApproveEntry[]>("list_auto_approve");
}

/** Append an entry. Backend validates `pattern` as a glob — bad globs reject. */
export async function addAutoApprove(entry: AutoApproveEntry): Promise<void> {
  return invoke("add_auto_approve", {
    tool: entry.tool,
    pattern: entry.pattern,
    profile: entry.profile ?? null,
  });
}

/** Remove the entry at `index` (0-based against `listAutoApprove()`). */
export async function removeAutoApprove(index: number): Promise<void> {
  return invoke("remove_auto_approve", { index });
}

/**
 * Best-effort guess of the natural glob pattern for a tool-call payload.
 * Mirrors the backend's `payload_for_match` so the "Always allow" button
 * suggests a pattern the user can immediately read and tweak.
 */
export function guessAutoApprovePattern(payload: unknown): string {
  if (payload && typeof payload === "object") {
    const obj = payload as Record<string, unknown>;
    for (const key of ["command", "cmd", "shell", "bash", "path", "file"]) {
      const v = obj[key];
      if (typeof v === "string" && v.trim()) {
        // First whitespace-separated token + `*` is the usual "let through
        // this family of calls" suggestion (e.g. `git status*`).
        const head = v.trim().split(/\s+/)[0];
        return head ? `${head}*` : v.trim();
      }
    }
  }
  if (typeof payload === "string" && payload.trim()) return payload.trim();
  return "*";
}

File: src-tauri/src/commands/approvals.rs
//! Tauri commands for managing persistent approval rules.

use std::path::PathBuf;

use crate::orchestrator::{
    load_policy, write_policy, ApprovalPolicy, ApprovalRules, AutoApproveEntry, AutoApproveList,
    Decision,
};

/// Append a new approval rule to `<project_root>/.cortex/approvals.toml`.
/// The `decision` arg accepts `"approve"` / `"allow"` and `"deny"` /
/// `"reject"`. Errors bubble up as user-visible strings.
#[tauri::command]
pub async fn add_approval_rule(
    project_root: String,
    pattern: String,
    decision: String,
) -> Result<(), String> {
    if project_root.trim().is_empty() {
        return Err("project_root is required".into());
    }
    if pattern.trim().is_empty() {
        return Err("pattern is required".into());
    }
    let parsed_decision = Decision::parse(&decision)
        .ok_or_else(|| format!("invalid decision '{decision}' (use 'approve' or 'deny')"))?;
    let root = PathBuf::from(&project_root);
    ApprovalRules::append_rule(&root, &pattern, parsed_decision).map_err(|e| e.to_string())
}

/// Return the contents of `~/.cortex/auto-approve.json` (or an empty list).
#[tauri::command]
pub async fn list_auto_approve() -> Result<Vec<AutoApproveEntry>, String> {
    Ok(AutoApproveList::list())
}

/// Append a `{tool, pattern, profile?}` entry to the global allowlist.
/// `pattern` is validated as a glob (`globset`). Duplicates are silently
/// de-duped.
#[tauri::command]
pub async fn add_auto_approve(
    tool: String,
    pattern: String,
    profile: Option<String>,
) -> Result<(), String> {
    if pattern.trim().is_empty() {
        return Err("pattern is required".into());
    }
    let entry = AutoApproveEntry {
        tool: tool.trim().to_string(),
        pattern: pattern.trim().to_string(),
        profile: profile.and_then(|p| {
            let t = p.trim().to_string();
            if t.is_empty() { None } else { Some(t) }
        }),
    };
    AutoApproveList::add(entry).map_err(|e| e.to_string())
}

/// Remove the entry at `index` (0-based against the current `list_auto_approve`
/// order). Out-of-range indices are a no-op.
#[tauri::command]
pub async fn remove_auto_approve(index: usize) -> Result<(), String> {
    AutoApproveList::remove(index).map_err(|e| e.to_string())
}

/// Read the configured approval policy for `project_root` (Codex-style "when do
/// we pause to ask?" axis). Missing / malformed files yield the default
/// (`on-request`). Returns the kebab-case string.
#[tauri::command]
pub async fn get_approval_policy(project_root: String) -> Result<String, String> {
    if project_root.trim().is_empty() {
        return Err("project_root is required".into());
    }
    let root = PathBuf::from(&project_root);
    Ok(load_policy(&root).as_str().to_string())
}

/// Persist the approval policy for `project_root`. Accepted values:
/// `"untrusted"`, `"on-request"`, `"never"` (plus snake/camel variants — see
/// `ApprovalPolicy::parse`).
#[tauri::command]
pub async fn set_approval_policy(project_root: String, policy: String) -> Result<(), String> {
    if project_root.trim().is_empty() {
        return Err("project_root is required".into());
    }
    let parsed = ApprovalPolicy::parse(&policy)
        .ok_or_else(|| format!("invalid policy '{policy}' (use untrusted|on-request|never)"))?;
    let root = PathBuf::from(&project_root);
    write_policy(&root, parsed).map_err(|e| e.to_string())
}
